chainerx_cc/chainerx/routines/reduction.cc

```diff
@@ -98,4 +98,56 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
     return x_cast - LogSumExp(x_cast, axis.has_value() ? axis : OptionalAxes{1}, true);
 }
 
+Array Cumsum(const Array& a, absl::optional<int8_t> axis) {
+    int8_t axis_norm;
+    Array a_reshaped{};
+    if (axis.has_value()) {
+        axis_norm = internal::NormalizeAxis(*axis, a.ndim());
+        a_reshaped = Copy(a);
+    } else {
+        axis_norm = 0;
+        // TODO(imanishi): Fix after chainerx::Ravel is supported.
+        a_reshaped = a.Reshape(Shape{a.GetTotalSize()});
+    }
+
+    Shape out_shape = a_reshaped.shape();
+    Array out = Empty(out_shape, a_reshaped.dtype(), a_reshaped.device());
+    // Decide the output dtype for integral input dtype.
+    Dtype out_dtype{};
+    switch (GetKind(a_reshaped.dtype())) {
+        case DtypeKind::kBool:
+        case DtypeKind::kInt:  // fallthrough
+            out_dtype = Dtype::kInt64;
+            break;
+        case DtypeKind::kUInt:
+            out_dtype = Dtype::kInt64;  // TODO(niboshi): This should be kUInt64
+            break;
+        default:
+            out_dtype = a_reshaped.dtype();
+    }
+
+    const Array& out_cast = out.AsType(out_dtype);
+    const Array& a_cast = a_reshaped.dtype() != out_cast.dtype() ? a_reshaped.AsType(out_cast.dtype()) : a_reshaped;
+
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<CumsumKernel>(a_cast, axis_norm, out_cast);
+    }
+
+    BackwardBuilder bb{"cumsum", a, out_cast};
```

Comment:
For simplification, it seems better to give `a_cast` (or `a_reshaped`) as an argument.

Reply:
I had originally given `a_reshaped` as an argument to `bb`, but we require `in_shape` to be `a_shape` and not `a_reshaped` since the shape of the gradient vector must be same as the input.

Reply:
> since the shape of the gradient vector must be same as the input.
Yes. the shape of the gradient vector (`input_grad`) must be the same as the shape of input (an argument of `bb`). So you should fix also backward implementation.

Reply:
Sorry, I am slightly confused. If I change to `BackwardBuilder bb{"cumsum", a_reshaped, out}`, then `input_grad` will have to be the same shape as `a_reshaped` right (Even if I modify the backward implementation)? So, in the case where the `axis=None`, `a_reshaped` will have a different shape than `a`.

Reply:
Done.